src/scala/com/github/johnynek/bazel_deps/IO.scala

```diff
@@ -115,83 +111,194 @@ object IO {
         System.exit(0)
     }
 
+  def orConst[A](optIO: Option[Result[A]], a: A): Result[A] =
+    optIO match {
+      case Some(io) => io
+      case None => IO.const(a)
+    }
+
+  def orUnit(optIO: Option[Result[Unit]]): Result[Unit] =
+    orConst(optIO, ())
+
   def fileSystemExec(root: File): FunctionK[Ops, Try] =
-    new FunctionK[Ops, Try] {
-      require(root.isAbsolute, s"Absolute path required, found: $root")
-
-      def fileFor(p: Path): File =
-        p.parts.foldLeft(root) { (p, element) => new File(p, element) }
-
-      def apply[A](o: Ops[A]): Try[A] = o match {
-        case Exists(f) => Try(fileFor(f).exists())
-        case MkDirs(f) => Try(fileFor(f).mkdirs())
-        case RmRf(f, removeHidden) =>
-          Try {
-            // get the java path
-            val file = fileFor(f)
-            // require(file.isDirectory, s"$f is not a directory")
-            val path = file.toPath
-            if (!removeHidden) {
-              Files.walkFileTree(
-                path,
-                new SimpleFileVisitor[JPath] {
-                  override def visitFile(
-                      file: JPath,
-                      attrs: BasicFileAttributes
-                  ) = {
-                    if (file.getFileName.startsWith(".") && !removeHidden) { // Hidden!
-                      throw new Exception(
-                        s"Encountered hidden file ${file.getFileName}, and should not remove hidden files/folders. Aborting."
-                      )
-                    }
-                    FileVisitResult.CONTINUE
-                  }
-                }
-              )
-            }
+    new ReadWriteExec(root)
+
+  class ReadOnlyExec(root: File) extends FunctionK[Ops, Try] {
+    require(root.isAbsolute, s"Absolute path required, found: $root")
+
+    protected val successUnit: Try[Unit] = Success(())    
 
+    def fileFor(p: Path): File =
+      p.parts.foldLeft(root) { (p, element) => new File(p, element) }
+
+    def apply[A](o: Ops[A]): Try[A] = o match {
+      case Exists(f) => Try(fileFor(f).exists())
+      case ReadFile(f) =>
+        Try({
+          val ff = fileFor(f)
+          if (ff.exists)
+            Some(new String(Files.readAllBytes(ff.toPath), charset))
+          else None
+        })
+      case Failed(err) => Failure(err)
+      // The following are not read only
+      case MkDirs(f) => Try(!fileFor(f).exists())
```

Comment:
So in a read only case this is kinda overloading the `MkDirs` op to check for directory existence?

Reply:
mkdirs has to return if it made the directory or not. We aren't making it, but we *would* make it in a phantom write situation if the directory is absent.

That said, I don't think we need to get cute. I think we can just fail all write operations in read-only mode.